### PR TITLE
Add peer-io-adapter

### PR DESCRIPTION
**Pull request by quentinl-c**
Features added :
* Client-side : the editor sends logoot operations and user informations through a p2p network without any assitance from server (using web-rtc technology)
* Server-side : server manages new client connections to the document and give a list of peers already connected (and other informations)

The new feature is accessible by this url : http://host/peer/:docID

Elements added : 
* Server-side : 
    * peer-io-adapter
* client-side : 
    * peer-io-adapter
    * peerjs library

In order to use web-rtc technology, the application deals with peerJS API : http://peerjs.com/, and actually using external signaling server (with 50 current connections limitation).

A signaling  server is set up  on expressJS application, but some troubles were encountered with firefox (singaling server doesn't forward offers from firefox client, when a firefox client tries to join a document created by a chrome client)

### CODE REVIEW

File: lib/coordinator.js [hunk 1]
```diff
@@ -30,568 +30,594 @@ var LogootSAdd = require('mute-structs').LogootSAdd;
...
+        }
+    };
+
+    this.serverDB.models.query()
+    .filter('docID', this.docID)
+    .execute()
+    .done(function (results) {
+        var doc;
+        if(results.length === 0) {
+            console.log("BLOB");
```
Comment (MatthieuNICOLAS) [inline lib/coordinator.js]: Remove this console.log

File: lib/coordinator.js [hunk 2]
```diff
@@ -30,568 +30,594 @@ var LogootSAdd = require('mute-structs').LogootSAdd;
...
+                bufferLocalLogootSOp: coordinator.bufferLocalLogootSOp,
+                mode: ONLINE_MODE,
+            };
+            coordinator.serverDB.models.add(doc)
+            .done(function (item) {
+                callback();
+            });
+        }
+        else {
+            console.log("BLOG");
```
Comment (MatthieuNICOLAS) [inline lib/coordinator.js]: Same as before

File: lib/coordinator.js [hunk 3]
```diff
@@ -30,568 +30,594 @@ var LogootSAdd = require('mute-structs').LogootSAdd;
...
+    this.replicaNumber = json.replicaNumber;
+    coordinator.updateLastModificationDate(new Date(json.lastModificationDate).valueOf());
+    this.creationDate = json.creationDate;
+
+    if(this.editor !== null) {
+        this.editor.removeListener('localOperation', this.localOperationHandler);
+    }
+
+    if(json.ropes !== null && json.ropes !== undefined) {
+        console.log("bobo");
```
Comment (MatthieuNICOLAS) [inline lib/coordinator.js]: Remove this line

File: lib/peer-io-adapter.js [hunk 4]
```diff
@@ -0,0 +1,445 @@
...
+var Data = function(event, data){
+    this.event = event;
+    this.data = data;
+};
+
+/*
+PeerInfo Object
+Inofmrations about remote peer
+connection : connection between the current and the remote peer
+relpicaNumber : replica number of the remote peer
```
Comment (MatthieuNICOLAS) [inline lib/peer-io-adapter.js]: Fix spelling mistakes

File: lib/peer-io-adapter.js [hunk 5]
```diff
@@ -0,0 +1,445 @@
...
+        }
+    });
+
+    this.coordinator.on('coordinatorDisposed', function (data) {
+        if(peerIOAdapter.disposed === false) {
+            peerIOAdapter.onCoordinatorDisposedHandler(data);
+        }
+    });
+    this.coordinator.on('doc', function (args){
+        //Give a copy of the
```
Comment (MatthieuNICOLAS) [inline lib/peer-io-adapter.js]: Complete this sentence

File: lib/peer-io-adapter.js [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+    this.coordinator.on('coordinatorDisposed', function (data) {
+        if(peerIOAdapter.disposed === false) {
+            peerIOAdapter.onCoordinatorDisposedHandler(data);
+        }
+    });
+    this.coordinator.on('doc', function (args){
+        //Give a copy of the 
+        console.log(args);
+        msg = {
+            "ropes": args.ropes,
```
Comment (MatthieuNICOLAS) [inline lib/peer-io-adapter.js]: Remove these "

File: lib/peer-io-adapter.js [hunk 7]
```diff
@@ -0,0 +1,445 @@
...
+        if(connection !== null){
+            connection.send(data);
+        }
+
+    });
+
+
+    this.coordinator.on('initDoc', function (args){
+        msg = {
+            "ropes": args.ropes,
```
Comment (MatthieuNICOLAS) [inline lib/peer-io-adapter.js]: Remove these "

File: lib/peer-io-adapter.js [hunk 8]
```diff
@@ -0,0 +1,445 @@
...
+
+PeerIOAdapter.prototype.createSocket = function () {
+    var peerIOAdapter = this;
+    this.peers = [];// empty the remote peers list
+    var connOptions = {
+        'sync disconnect on unload': true
+    };
+
+    this.socketServer = io.connect(location.origin, connOptions);
+    //var peerServerId = Math.floor(Math.random()*100000).toString();
```
Comment (MatthieuNICOLAS) [inline lib/peer-io-adapter.js]: Document these lines

File: lib/peer-io-adapter.js [hunk 9]
```diff
@@ -0,0 +1,445 @@
...
+    function handleEvent(args){
+        //manage data receiving 
+        console.log("HandleEvent");
+        console.log(args);
+        try{
+            var data = JSON.parse(args);
+            if(data.event !== null && data.event !== undefined){
+                switch(data.event){
+                    case 'sendOps':
+                        if(peerIOAdapter.disposed === false) {
```
Comment (MatthieuNICOLAS) [inline lib/peer-io-adapter.js]: Refactor this condition

File: lib/peer-io-adapter.js [hunk 10]
```diff
@@ -0,0 +1,445 @@
...
+                            peerIOAdapter.emit('receiveParole', data.data);
+                        }
+                        break;
+                    case 'connect' :
+                        if(peerIOAdapter.disposed === false) {
+                            peerIOAdapter.emit('connect');
+                        }
+                        break;
+                    case 'disconnect' :
```
Comment (MatthieuNICOLAS) [inline lib/peer-io-adapter.js]: Remove this case

File: lib/socket-io-adapter.js [hunk 11]
```diff
@@ -59,6 +59,8 @@ SocketIOAdapter.prototype.setInfosUsersModule = function (infosUsersModule) {
     this.infosUsersModule = infosUsersModule;
 
     infosUsersModule.on('changeLocalCursorAndSelections', function (data) {
+        console.log("plop");
```
Comment (MatthieuNICOLAS) [inline lib/socket-io-adapter.js]: Remove theses lines

File: lib/coordinator.js (hunk 1 above)
Comment (quentinl-c) [inline lib/coordinator.js]: done

File: lib/coordinator.js (hunk 2 above)
Comment (quentinl-c) [inline lib/coordinator.js]: done

File: lib/coordinator.js (hunk 3 above)
Comment (quentinl-c) [inline lib/coordinator.js]: done

File: lib/peer-io-adapter.js (hunk 4 above)
Comment (quentinl-c) [inline lib/peer-io-adapter.js]: done

File: lib/peer-io-adapter.js (hunk 5 above)
Comment (quentinl-c) [inline lib/peer-io-adapter.js]: done

File: lib/peer-io-adapter.js (hunk 6 above)
Comment (quentinl-c) [inline lib/peer-io-adapter.js]: done

File: lib/peer-io-adapter.js (hunk 7 above)
Comment (quentinl-c) [inline lib/peer-io-adapter.js]: done

File: lib/peer-io-adapter.js (hunk 8 above)
Comment (quentinl-c) [inline lib/peer-io-adapter.js]: done

File: lib/peer-io-adapter.js (hunk 9 above)
Comment (quentinl-c) [inline lib/peer-io-adapter.js]: done

File: lib/peer-io-adapter.js (hunk 10 above)
Comment (quentinl-c) [inline lib/peer-io-adapter.js]: done

File: lib/socket-io-adapter.js (hunk 11 above)
Comment (quentinl-c) [inline lib/socket-io-adapter.js]: done

**Pull request merged**